{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:10.154465Z",
     "start_time": "2026-02-13T11:30:08.003018Z"
    }
   },
   "source": "!pip install pandas-plink limix-lmm",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas-plink in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (2.3.2)\r\n",
      "Requirement already satisfied: limix-lmm in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (0.1.2)\r\n",
      "Requirement already satisfied: cffi>=1.17 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (2.0.0)\r\n",
      "Requirement already satisfied: dask[dataframe] in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (2026.1.2)\r\n",
      "Requirement already satisfied: deprecated<2.0,>=1.2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (1.3.1)\r\n",
      "Requirement already satisfied: numpy<2.0,>=1.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (1.26.4)\r\n",
      "Requirement already satisfied: pandas<3.0,>=2.2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (2.3.3)\r\n",
      "Requirement already satisfied: pandera[dask,mypy] in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (0.29.0)\r\n",
      "Requirement already satisfied: tqdm<5.0,>=4.67 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (4.67.3)\r\n",
      "Requirement already satisfied: xarray in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (2025.6.1)\r\n",
      "Requirement already satisfied: zstandard<0.24,>=0.23 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-plink) (0.23.0)\r\n",
      "Requirement already satisfied: wrapt<3,>=1.10 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from deprecated<2.0,>=1.2->pandas-plink) (2.1.1)\r\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas<3.0,>=2.2->pandas-plink) (2.9.0.post0)\r\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas<3.0,>=2.2->pandas-plink) (2025.2)\r\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas<3.0,>=2.2->pandas-plink) (2025.3)\r\n",
      "Requirement already satisfied: scipy>=1.1.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from limix-lmm) (1.15.3)\r\n",
      "Requirement already satisfied: pytest>=3.6.3 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from limix-lmm) (9.0.2)\r\n",
      "Requirement already satisfied: limix-core>=1.0.2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from limix-lmm) (1.0.2)\r\n",
      "Requirement already satisfied: matplotlib>=2.2.3 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from limix-lmm) (3.10.8)\r\n",
      "Requirement already satisfied: bgen-reader>=2.0.7 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from limix-lmm) (4.0.8)\r\n",
      "Requirement already satisfied: geno-sugar in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from limix-lmm) (0.1.1)\r\n",
      "Requirement already satisfied: appdirs>=1.4.3 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (1.4.4)\r\n",
      "Requirement already satisfied: cachetools>=3.1.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (7.0.1)\r\n",
      "Requirement already satisfied: cbgen>=1.0.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (1.0.5)\r\n",
      "Requirement already satisfied: requests>=1.25.8 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (2.32.5)\r\n",
      "Requirement already satisfied: texttable>=1.6.2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (1.7.0)\r\n",
      "Requirement already satisfied: pooch in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from cbgen>=1.0.1->bgen-reader>=2.0.7->limix-lmm) (1.9.0)\r\n",
      "Requirement already satisfied: urllib3>=1.26 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from cbgen>=1.0.1->bgen-reader>=2.0.7->limix-lmm) (2.6.3)\r\n",
      "Requirement already satisfied: pycparser in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from cffi>=1.17->pandas-plink) (3.0)\r\n",
      "Requirement already satisfied: click>=8.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (8.3.1)\r\n",
      "Requirement already satisfied: cloudpickle>=3.0.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (3.1.2)\r\n",
      "Requirement already satisfied: fsspec>=2021.09.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (2026.2.0)\r\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (26.0)\r\n",
      "Requirement already satisfied: partd>=1.4.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (1.4.2)\r\n",
      "Requirement already satisfied: pyyaml>=5.3.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (6.0.3)\r\n",
      "Requirement already satisfied: toolz>=0.12.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (1.1.0)\r\n",
      "Requirement already satisfied: importlib_metadata>=4.13.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (8.7.1)\r\n",
      "Requirement already satisfied: pyarrow>=16.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from dask[array,bag,dataframe,delayed]>=2.12.0->bgen-reader>=2.0.7->limix-lmm) (23.0.0)\r\n",
      "Requirement already satisfied: zipp>=3.20 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from importlib_metadata>=4.13.0->dask[dataframe]->pandas-plink) (3.23.0)\r\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (1.3.2)\r\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (0.12.1)\r\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (4.61.1)\r\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (1.4.9)\r\n",
      "Requirement already satisfied: pillow>=8 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (12.1.1)\r\n",
      "Requirement already satisfied: pyparsing>=3 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (3.3.2)\r\n",
      "Requirement already satisfied: locket in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from partd>=1.4.0->dask[dataframe]->pandas-plink) (1.0.0)\r\n",
      "Requirement already satisfied: exceptiongroup>=1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (1.3.1)\r\n",
      "Requirement already satisfied: iniconfig>=1.0.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (2.3.0)\r\n",
      "Requirement already satisfied: pluggy<2,>=1.5 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (1.6.0)\r\n",
      "Requirement already satisfied: pygments>=2.7.2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (2.19.2)\r\n",
      "Requirement already satisfied: tomli>=1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (2.4.0)\r\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from exceptiongroup>=1->pytest>=3.6.3->limix-lmm) (4.15.0)\r\n",
      "Requirement already satisfied: six>=1.5 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas<3.0,>=2.2->pandas-plink) (1.17.0)\r\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from requests>=1.25.8->bgen-reader>=2.0.7->limix-lmm) (3.4.4)\r\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from requests>=1.25.8->bgen-reader>=2.0.7->limix-lmm) (3.11)\r\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from requests>=1.25.8->bgen-reader>=2.0.7->limix-lmm) (2026.1.4)\r\n",
      "Requirement already satisfied: scikit-learn>=0.20.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from geno-sugar->limix-lmm) (1.7.2)\r\n",
      "Requirement already satisfied: joblib>=1.2.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from scikit-learn>=0.20.0->geno-sugar->limix-lmm) (1.5.3)\r\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from scikit-learn>=0.20.0->geno-sugar->limix-lmm) (3.6.0)\r\n",
      "Requirement already satisfied: pydantic in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (2.12.5)\r\n",
      "Requirement already satisfied: typeguard in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (4.4.4)\r\n",
      "Requirement already satisfied: typing_inspect>=0.6.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (0.9.0)\r\n",
      "Requirement already satisfied: pandas-stubs in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (2.3.3.260113)\r\n",
      "Requirement already satisfied: scipy-stubs in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (1.15.3.0)\r\n",
      "Requirement already satisfied: distributed in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (2026.1.2)\r\n",
      "Requirement already satisfied: mypy-extensions>=0.3.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from typing_inspect>=0.6.0->pandera[dask,mypy]->pandas-plink) (1.1.0)\r\n",
      "Requirement already satisfied: jinja2>=2.10.3 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (3.1.6)\r\n",
      "Requirement already satisfied: msgpack>=1.0.2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (1.1.2)\r\n",
      "Requirement already satisfied: psutil>=5.8.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (7.2.2)\r\n",
      "Requirement already satisfied: sortedcontainers>=2.0.5 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (2.4.0)\r\n",
      "Requirement already satisfied: tblib!=3.2.0,!=3.2.1,>=1.6.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (3.2.2)\r\n",
      "Requirement already satisfied: tornado>=6.2.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (6.5.4)\r\n",
      "Requirement already satisfied: zict>=3.0.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (3.0.0)\r\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from jinja2>=2.10.3->distributed->pandera[dask,mypy]->pandas-plink) (3.0.3)\r\n",
      "Requirement already satisfied: types-pytz>=2022.1.1 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pandas-stubs->pandera[dask,mypy]->pandas-plink) (2025.2.0.20251108)\r\n",
      "Requirement already satisfied: platformdirs>=2.5.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pooch->cbgen>=1.0.1->bgen-reader>=2.0.7->limix-lmm) (4.5.1)\r\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pydantic->pandera[dask,mypy]->pandas-plink) (0.7.0)\r\n",
      "Requirement already satisfied: pydantic-core==2.41.5 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pydantic->pandera[dask,mypy]->pandas-plink) (2.41.5)\r\n",
      "Requirement already satisfied: typing-inspection>=0.4.2 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from pydantic->pandera[dask,mypy]->pandas-plink) (0.4.2)\r\n",
      "Requirement already satisfied: optype>=0.9.3 in /Users/malenavoss/GoBi_Theis/GWAS Blockteil/gobi_gwas/venv/lib/python3.10/site-packages (from scipy-stubs->pandera[dask,mypy]->pandas-plink) (0.9.3)\r\n",
      "\r\n",
      "\u001B[1m[\u001B[0m\u001B[34;49mnotice\u001B[0m\u001B[1;39;49m]\u001B[0m\u001B[39;49m A new release of pip is available: \u001B[0m\u001B[31;49m26.0\u001B[0m\u001B[39;49m -> \u001B[0m\u001B[32;49m26.0.1\u001B[0m\r\n",
      "\u001B[1m[\u001B[0m\u001B[34;49mnotice\u001B[0m\u001B[1;39;49m]\u001B[0m\u001B[39;49m To update, run: \u001B[0m\u001B[32;49mpip install --upgrade pip\u001B[0m\r\n"
     ]
    }
   ],
   "execution_count": 170
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:10.171551Z",
     "start_time": "2026-02-13T11:30:10.161008Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import pdb\n",
    "from pandas_plink import read_plink\n",
    "from limix_lmm import LMM\n",
    "import pylab as plt\n",
    "import os\n",
    "import pandas as pd"
   ],
   "id": "4bdc1c8fbeabb6d5",
   "outputs": [],
   "execution_count": 171
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:10.190116Z",
     "start_time": "2026-02-13T11:30:10.176980Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import scipy\n",
    "import scipy.stats as st\n",
    "\n",
    "if not hasattr(scipy, 'dot'):\n",
    "    scipy.dot = np.dot\n",
    "if not hasattr(scipy, 'einsum'):\n",
    "    scipy.einsum = np.einsum\n",
    "if not hasattr(scipy, 'log'):\n",
    "    scipy.log = np.log\n",
    "if not hasattr(scipy, 'sign'):\n",
    "    scipy.sign = np.sign\n",
    "if not hasattr(scipy, 'sqrt'):\n",
    "    scipy.sqrt = np.sqrt"
   ],
   "id": "30cfb5acfb05c839",
   "outputs": [],
   "execution_count": 172
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:14.835004Z",
     "start_time": "2026-02-13T11:30:10.193117Z"
    }
   },
   "cell_type": "code",
   "source": [
    "!wget -O ALL.chr22_GRCh38.genotypes.20170504.zip \"https://www.dropbox.com/s/86fdj3i7ir3vap7/ALL.chr22_GRCh38.genotypes.20170504.zip?dl=1\"\n",
    "!unzip -o -q ALL.chr22_GRCh38.genotypes.20170504.zip && rm ALL.chr22_GRCh38.genotypes.20170504.zip"
   ],
   "id": "d7a94518b1543084",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2026-02-13 12:30:10--  https://www.dropbox.com/s/86fdj3i7ir3vap7/ALL.chr22_GRCh38.genotypes.20170504.zip?dl=1\r\n",
      "Resolving www.dropbox.com (www.dropbox.com)... 2620:100:6022:18::a27d:4212, 162.125.66.18\r\n",
      "Connecting to www.dropbox.com (www.dropbox.com)|2620:100:6022:18::a27d:4212|:443... connected.\r\n",
      "HTTP request sent, awaiting response... 302 Found\r\n",
      "Location: https://www.dropbox.com/scl/fi/2j079oew31gwd2hzajlil/ALL.chr22_GRCh38.genotypes.20170504.zip?rlkey=7307x4sjq9lkvhwu3aivh5uut&dl=1 [following]\r\n",
      "--2026-02-13 12:30:11--  https://www.dropbox.com/scl/fi/2j079oew31gwd2hzajlil/ALL.chr22_GRCh38.genotypes.20170504.zip?rlkey=7307x4sjq9lkvhwu3aivh5uut&dl=1\r\n",
      "Reusing existing connection to [www.dropbox.com]:443.\r\n",
      "HTTP request sent, awaiting response... 302 Found\r\n",
      "Location: https://uc7bd928322a5014c6c57c52bbbf.dl.dropboxusercontent.com/cd/0/inline/C61QmUeVdktmHKvFQ0MmbPjZmxFhZXRTkJ0MTxs4TQlWzUWf96ij8MkjQKSqkPy_EsBxxKcrKnR_YsdU0bZDv5D39o_x8QoeRLXkCzfWifBQV642vYcMppDq6pGnfyQotXo/file?dl=1# [following]\r\n",
      "--2026-02-13 12:30:11--  https://uc7bd928322a5014c6c57c52bbbf.dl.dropboxusercontent.com/cd/0/inline/C61QmUeVdktmHKvFQ0MmbPjZmxFhZXRTkJ0MTxs4TQlWzUWf96ij8MkjQKSqkPy_EsBxxKcrKnR_YsdU0bZDv5D39o_x8QoeRLXkCzfWifBQV642vYcMppDq6pGnfyQotXo/file?dl=1\r\n",
      "Resolving uc7bd928322a5014c6c57c52bbbf.dl.dropboxusercontent.com (uc7bd928322a5014c6c57c52bbbf.dl.dropboxusercontent.com)... 2620:100:6022:15::a27d:420f, 162.125.66.15\r\n",
      "Connecting to uc7bd928322a5014c6c57c52bbbf.dl.dropboxusercontent.com (uc7bd928322a5014c6c57c52bbbf.dl.dropboxusercontent.com)|2620:100:6022:15::a27d:420f|:443... connected.\r\n",
      "HTTP request sent, awaiting response... 302 Found\r\n",
      "Location: /cd/0/inline2/C61ITIOggrZ2ZPQXIWLC9xRoPV9rtyKc1k5tDYyW-1HKZyNCMSNCbHAK6NCtYl-ssOfqYp450RJnMuR2PJuFMuFNCle0WIyJjmn8GhWEhOCoUPAU3SQy7F20bG2FCsoTpGMOfs7TtVpfRlJViqwAighmyETtTiEXn0kZY18gYjE9zK9OVuA2I0kyCDdk5Q5FPbX9Ak2C0BOb25a0qe8_GLDiW9EUpI1BeTTYsiiyiIZclJhXz55rxxiv_6kUKlErlphainBP0lz1lZmg5_u6onWahMFzvVk5o2YiJzQz1aXgmAK6DKvQs6loiRLOJGejA098poDcdSpR2-Hrenk7mL9bOgA2Fyfvgq9Hxgz9C15H2w/file?dl=1 [following]\r\n",
      "--2026-02-13 12:30:12--  https://uc7bd928322a5014c6c57c52bbbf.dl.dropboxusercontent.com/cd/0/inline2/C61ITIOggrZ2ZPQXIWLC9xRoPV9rtyKc1k5tDYyW-1HKZyNCMSNCbHAK6NCtYl-ssOfqYp450RJnMuR2PJuFMuFNCle0WIyJjmn8GhWEhOCoUPAU3SQy7F20bG2FCsoTpGMOfs7TtVpfRlJViqwAighmyETtTiEXn0kZY18gYjE9zK9OVuA2I0kyCDdk5Q5FPbX9Ak2C0BOb25a0qe8_GLDiW9EUpI1BeTTYsiiyiIZclJhXz55rxxiv_6kUKlErlphainBP0lz1lZmg5_u6onWahMFzvVk5o2YiJzQz1aXgmAK6DKvQs6loiRLOJGejA098poDcdSpR2-Hrenk7mL9bOgA2Fyfvgq9Hxgz9C15H2w/file?dl=1\r\n",
      "Reusing existing connection to [uc7bd928322a5014c6c57c52bbbf.dl.dropboxusercontent.com]:443.\r\n",
      "HTTP request sent, awaiting response... 200 OK\r\n",
      "Length: 18535586 (18M) [application/binary]\r\n",
      "Saving to: ‘ALL.chr22_GRCh38.genotypes.20170504.zip’\r\n",
      "\r\n",
      "ALL.chr22_GRCh38.ge 100%[===================>]  17.68M  19.1MB/s    in 0.9s    \r\n",
      "\r\n",
      "2026-02-13 12:30:14 (19.1 MB/s) - ‘ALL.chr22_GRCh38.genotypes.20170504.zip’ saved [18535586/18535586]\r\n",
      "\r\n"
     ]
    }
   ],
   "execution_count": 173
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:15.389071Z",
     "start_time": "2026-02-13T11:30:14.842812Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# load genetic data\n",
    "bfile = 'ALL.chr22_GRCh38.genotypes.20170504'\n",
    "bim, fam, G = read_plink(bfile)"
   ],
   "id": "7e2db59f541fe205",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00,  8.10it/s]\n"
     ]
    }
   ],
   "execution_count": 174
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Preprocessing",
   "id": "a6fe56a950ff9247"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:15.404324Z",
     "start_time": "2026-02-13T11:30:15.392266Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(G.shape)   # (n_variants, n_samples)\n",
    "print(bim.head())\n",
    "print(fam.head())\n",
    "print(G.shape[0], bim.shape[0], fam.shape[0])\n"
   ],
   "id": "92fe8ccb31524212",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(109827, 2504)\n",
      "  chrom          snp   cm       pos a0 a1  i\n",
      "0    22    rs9617549  0.0  10874444  T  C  0\n",
      "1    22    rs5747224  0.0  11122151  A  G  1\n",
      "2    22    rs2456393  0.0  11122417  T  G  2\n",
      "3    22  rs199668908  0.0  11123683  C  T  3\n",
      "4    22    rs4010319  0.0  11123736  C  T  4\n",
      "       fid      iid father mother gender trait  i\n",
      "0  HG00096  HG00096      0      0      0    -9  0\n",
      "1  HG00097  HG00097      0      0      0    -9  1\n",
      "2  HG00099  HG00099      0      0      0    -9  2\n",
      "3  HG00100  HG00100      0      0      0    -9  3\n",
      "4  HG00101  HG00101      0      0      0    -9  4\n",
      "109827 109827 2504\n"
     ]
    }
   ],
   "execution_count": 175
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:16.239554Z",
     "start_time": "2026-02-13T11:30:15.406359Z"
    }
   },
   "cell_type": "code",
   "source": [
    "#remove variants with many missing data\n",
    "snp_missing = np.mean(np.isnan(G), axis=1).compute()\n",
    "keep_missing = snp_missing < 0.01\n",
    "\n",
    "G = G[keep_missing, :]\n",
    "bim = bim.iloc[keep_missing].reset_index(drop=True)"
   ],
   "id": "b4ba6e2a6f8d2246",
   "outputs": [],
   "execution_count": 176
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:16.249084Z",
     "start_time": "2026-02-13T11:30:16.243413Z"
    }
   },
   "cell_type": "code",
   "source": "print(G.shape) ",
   "id": "e0238947a37c252e",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(109803, 2504)\n"
     ]
    }
   ],
   "execution_count": 177
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:17.350790Z",
     "start_time": "2026-02-13T11:30:16.254763Z"
    }
   },
   "cell_type": "code",
   "source": [
    "#removing duplicates of variants to avoid bias\n",
    "duplicate = bim.duplicated(subset=['chrom', 'pos'], keep = 'first')\n",
    "\n",
    "G.compute_chunk_sizes()\n",
    "G = G[~duplicate, :]\n",
    "bim = bim.loc[~duplicate].reset_index(drop=True)\n"
   ],
   "id": "4fb75118b964c15f",
   "outputs": [],
   "execution_count": 178
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:17.359195Z",
     "start_time": "2026-02-13T11:30:17.354729Z"
    }
   },
   "cell_type": "code",
   "source": "print(G.shape) ",
   "id": "ec9f60b7afc2faf5",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(109712, 2504)\n"
     ]
    }
   ],
   "execution_count": 179
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.273861Z",
     "start_time": "2026-02-13T11:30:17.363316Z"
    }
   },
   "cell_type": "code",
   "source": [
    "#retaining only common variants -MAF >=1% or 5%\n",
    "allele_freq = np.nanmean(G, axis=1)/2 \n",
    "maf = np.minimum(allele_freq, 1-allele_freq)\n",
    "\n",
    "keep_maf = maf > 0.05\n",
    "G = G[keep_maf, :]\n",
    "bim = bim.iloc[keep_maf].reset_index(drop=True)\n",
    "G.compute_chunk_sizes()  "
   ],
   "id": "d467c82048e3ec3e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dask.array<getitem_variadic, shape=(109712, 2504), dtype=float32, chunksize=(1016, 1024), chunktype=numpy.ndarray>"
      ],
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 1.02 GiB </td>\n",
       "                        <td> 3.97 MiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (109712, 2504) </td>\n",
       "                        <td> (1016, 1024) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 324 chunks in 768 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> float32 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"79\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"29\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"29\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"29\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"29\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"29\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"29\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"29\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"43\" x2=\"29\" y2=\"43\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"29\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"29\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"62\" x2=\"29\" y2=\"62\" />\n",
       "  <line x1=\"0\" y1=\"68\" x2=\"29\" y2=\"68\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"29\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"29\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"87\" x2=\"29\" y2=\"87\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"29\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"29\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"106\" x2=\"29\" y2=\"106\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"29\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"29\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"12\" y1=\"0\" x2=\"12\" y2=\"120\" />\n",
       "  <line x1=\"24\" y1=\"0\" x2=\"24\" y2=\"120\" />\n",
       "  <line x1=\"29\" y1=\"0\" x2=\"29\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 29.78274806141775,0.0 29.78274806141775,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"14.891374030708874\" y=\"140.0\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >2504</text>\n",
       "  <text x=\"49.78274806141775\" y=\"60.0\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,49.78274806141775,60.0)\">109712</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table>"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 180
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.279387Z",
     "start_time": "2026-02-13T11:30:23.275911Z"
    }
   },
   "cell_type": "code",
   "source": "print(G.shape) ",
   "id": "f70788f96dc8fc2e",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(109712, 2504)\n"
     ]
    }
   ],
   "execution_count": 181
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.286066Z",
     "start_time": "2026-02-13T11:30:23.281675Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def extract_region(bim, G, chrom, start, end):\n",
    "    I1 = bim['chrom']==str(chrom)\n",
    "    I2 = bim['pos'] > start\n",
    "    I3 = bim['pos'] < end\n",
    "    Ikeep = I1 * I2 * I3\n",
    "    bim = bim.loc[Ikeep].copy()\n",
    "    G = G[bim['i'].values]\n",
    "    bim['i'] = np.arange(len(bim))\n",
    "    return bim, G"
   ],
   "id": "cd8cd633a0ff45a",
   "outputs": [],
   "execution_count": 182
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.293205Z",
     "start_time": "2026-02-13T11:30:23.288559Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def subset_individuals(fam, G, keep_idxs):\n",
    "    fam = fam.iloc[keep_idxs]\n",
    "    G = G[:, keep_idxs]\n",
    "    return bim, G"
   ],
   "id": "2b0ee70a66975709",
   "outputs": [],
   "execution_count": 183
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.315603Z",
     "start_time": "2026-02-13T11:30:23.296880Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# subset to 1000 individuals\n",
    "N = 1000\n",
    "np.random.seed(0)\n",
    "idxs = np.sort(np.random.permutation(fam.shape[0])[:N])\n",
    "fams, Gs = subset_individuals(fam, G, idxs)"
   ],
   "id": "f4157a1d63bc16a3",
   "outputs": [],
   "execution_count": 184
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.340314Z",
     "start_time": "2026-02-13T11:30:23.318021Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# extract region\n",
    "chrom = 22\n",
    "start = 30797531\n",
    "size = 100000\n",
    "bimr, Gr = extract_region(bim, Gs, chrom, start, start + size)"
   ],
   "id": "16f4cb7ab38d6665",
   "outputs": [],
   "execution_count": 185
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.350274Z",
     "start_time": "2026-02-13T11:30:23.342746Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# load genotype matrix\n",
    "X = Gr.T"
   ],
   "id": "4909095f5ad3d2e3",
   "outputs": [],
   "execution_count": 186
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Phase 1: Phenotype Simulation",
   "id": "2dc818c940fcb168"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.362348Z",
     "start_time": "2026-02-13T11:30:23.354478Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def simulate_pheno(X, idx_caus, var_expl, direction=None):\n",
    "    # Ensure that the number of causal variant indices matches the number of variances explained.\n",
    "    assert len(idx_caus) == len(var_expl)\n",
    "\n",
    "    # If no direction is provided, randomly assign a positive or negative direction for each causal variant.\n",
    "    if direction is None:\n",
    "        direction = 2. * (np.random.rand(len(idx_caus)) > 0.5) - 1.\n",
    "    # Ensure that the number of directions matches the number of causal variant indices.\n",
    "    assert len(idx_caus) == len(direction)\n",
    "\n",
    "    # Compute the remaining variance after accounting for the variance explained by the causal variants.\n",
    "    ve = 1 - var_expl.sum()\n",
    "    # Ensure that the total variance explained by causal variants is less than 1.\n",
    "    assert ve > 0, 'sum(var_expl) should be < 1'\n",
    "\n",
    "    # Compute the effect sizes for the causal variants based on the variance they explain and their direction.\n",
    "    beta = np.sqrt(var_expl) * direction\n",
    "\n",
    "    # Extract the columns of X corresponding to the causal variants and standardize them.\n",
    "    Xc = X[:, idx_caus]\n",
    "    Xc = (Xc - Xc.mean(0)) / Xc.std(0)\n",
    "\n",
    "    # Compute the genetic component of the phenotype.\n",
    "    yg = Xc.dot(beta)[:, None]\n",
    "    # Compute the noise component of the phenotype.\n",
    "    yn = np.sqrt(ve) * np.random.randn(X.shape[0], 1)\n",
    "\n",
    "    # Sum the genetic and noise components to get the simulated phenotype.\n",
    "    y = yg + yn\n",
    "\n",
    "    # Initialize the real effect sizes for all variants in X as zeros.\n",
    "    beta_real = np.zeros(X.shape[1])\n",
    "    # Update the real effect sizes for the causal variants.\n",
    "    beta_real[idx_caus] = beta\n",
    "\n",
    "    # Standardize the phenotypic values to have mean 0 and standard deviation 1.\n",
    "    ystd = y.std()\n",
    "    y = (y - y.mean()) / ystd\n",
    "    # Adjust the real effect sizes accordingly after standardizing y.\n",
    "    beta_real = beta_real / ystd\n",
    "\n",
    "    return y, beta_real"
   ],
   "id": "6eaa0b0904d6ed89",
   "outputs": [],
   "execution_count": 187
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.373766Z",
     "start_time": "2026-02-13T11:30:23.366773Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def qq_plot(p_values, figsize, title): #parameters changed\n",
    "    # Sort p-values\n",
    "    observed = -np.log10(np.sort(p_values))\n",
    "    expected = -np.log10(np.arange(1, len(p_values) + 1) / (len(p_values) + 2))\n",
    "\n",
    "    # Create the QQ plot\n",
    "    if figsize is not None: #added\n",
    "      plt.figure(figsize=figsize) #added\n",
    "    plt.scatter(expected, observed, marker='.')\n",
    "    plt.plot([0, max(expected)], [0, max(expected)], color='red', linestyle='--')\n",
    "    plt.xlabel('Expected -log10(P-value)')\n",
    "    plt.ylabel('Observed -log10(P-value)')\n",
    "    plt.title(title)"
   ],
   "id": "a13c405367879b33",
   "outputs": [],
   "execution_count": 188
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.414949Z",
     "start_time": "2026-02-13T11:30:23.376841Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Generate phenotype\n",
    "#n_causal = 10 # number of causal variants\n",
    "#idx_caus = np.random.choice(X.shape[1], size=n_causal)\n",
    "#var_expl = np.repeat(0.06, n_causal)\n",
    "\n",
    "h2_values = [0.1,0.6]\n",
    "n_causal_values = [10, 100]\n",
    "\n",
    "for h2 in h2_values:\n",
    "    for n_causal in n_causal_values:\n",
    "        idx_caus = np.random.choice(X.shape[1], size=n_causal) \n",
    "        var_expl = np.full(n_causal, h2 / n_causal)\n",
    "        y, beta_real = simulate_pheno(X, idx_caus, var_expl)"
   ],
   "id": "aa784056277e9bcb",
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "a must be greater than 0 unless no samples are taken",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mValueError\u001B[0m                                Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[189], line 11\u001B[0m\n\u001B[1;32m      9\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m h2 \u001B[38;5;129;01min\u001B[39;00m h2_values:\n\u001B[1;32m     10\u001B[0m     \u001B[38;5;28;01mfor\u001B[39;00m n_causal \u001B[38;5;129;01min\u001B[39;00m n_causal_values:\n\u001B[0;32m---> 11\u001B[0m         idx_caus \u001B[38;5;241m=\u001B[39m \u001B[43mnp\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mrandom\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mchoice\u001B[49m\u001B[43m(\u001B[49m\u001B[43mX\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mshape\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;241;43m1\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43msize\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mn_causal\u001B[49m\u001B[43m)\u001B[49m \n\u001B[1;32m     12\u001B[0m         var_expl \u001B[38;5;241m=\u001B[39m np\u001B[38;5;241m.\u001B[39mfull(n_causal, h2 \u001B[38;5;241m/\u001B[39m n_causal)\n\u001B[1;32m     13\u001B[0m         y, beta_real \u001B[38;5;241m=\u001B[39m simulate_pheno(X, idx_caus, var_expl)\n",
      "File \u001B[0;32mnumpy/random/mtrand.pyx:945\u001B[0m, in \u001B[0;36mnumpy.random.mtrand.RandomState.choice\u001B[0;34m()\u001B[0m\n",
      "\u001B[0;31mValueError\u001B[0m: a must be greater than 0 unless no samples are taken"
     ]
    }
   ],
   "execution_count": 189
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.417323Z",
     "start_time": "2026-02-13T11:30:23.417075Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Perform genetic association testing\n",
    "F = np.ones_like(y)\n",
    "lmm = LMM(y, F)\n",
    "lmm.process(X)\n",
    "pv = lmm.getPv()\n",
    "beta = lmm.getBetaSNP()\n",
    "beta_ste = lmm.getBetaSNPste()"
   ],
   "id": "639c2f0f3d23621",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.421364Z",
     "start_time": "2026-02-13T11:30:23.420776Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# TODO: Create and show QQ plot\n",
    "qq_plot(pv, None, 'QQ plot')\n",
    "plt.show()"
   ],
   "id": "bfd4b62067afd860",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-13T11:30:23.423679Z",
     "start_time": "2026-02-13T11:30:23.423071Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Plot ground truth and results\n",
    "x = bimr['pos'].values\n",
    "plt.subplot(211)\n",
    "plt.title('Real effect size')\n",
    "plt.plot(x, beta_real, '.k')\n",
    "plt.ylabel('eff size')\n",
    "plt.subplot(212)\n",
    "plt.title('GWAS results')\n",
    "plt.plot(x, -np.log10(pv), '.k')\n",
    "plt.ylabel('-log$_{10}$ P')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ],
   "id": "15d8f52adce9cc0",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
