{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3e4718c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas-plink in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (2.3.2)\n",
      "Requirement already satisfied: limix-lmm in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (0.1.2)\n",
      "Requirement already satisfied: cffi>=1.17 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (1.17.1)\n",
      "Requirement already satisfied: dask[dataframe] in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (2024.11.2)\n",
      "Requirement already satisfied: deprecated<2.0,>=1.2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (1.3.1)\n",
      "Requirement already satisfied: numpy<2.0,>=1.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (1.26.4)\n",
      "Requirement already satisfied: pandas<3.0,>=2.2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (2.3.3)\n",
      "Requirement already satisfied: pandera[dask,mypy] in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (0.27.0)\n",
      "Requirement already satisfied: tqdm<5.0,>=4.67 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (4.67.1)\n",
      "Requirement already satisfied: xarray in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (2025.6.1)\n",
      "Requirement already satisfied: zstandard<0.24,>=0.23 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-plink) (0.23.0)\n",
      "Requirement already satisfied: wrapt<3,>=1.10 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from deprecated<2.0,>=1.2->pandas-plink) (1.17.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas<3.0,>=2.2->pandas-plink) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas<3.0,>=2.2->pandas-plink) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas<3.0,>=2.2->pandas-plink) (2025.2)\n",
      "Requirement already satisfied: scipy>=1.1.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from limix-lmm) (1.15.3)\n",
      "Requirement already satisfied: pytest>=3.6.3 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from limix-lmm) (9.0.1)\n",
      "Requirement already satisfied: limix-core>=1.0.2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from limix-lmm) (1.0.2)\n",
      "Requirement already satisfied: matplotlib>=2.2.3 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from limix-lmm) (3.10.7)\n",
      "Requirement already satisfied: bgen-reader>=2.0.7 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from limix-lmm) (4.0.8)\n",
      "Requirement already satisfied: geno-sugar in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from limix-lmm) (0.1.1)\n",
      "Requirement already satisfied: appdirs>=1.4.3 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (1.4.4)\n",
      "Requirement already satisfied: cachetools>=3.1.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (6.2.2)\n",
      "Requirement already satisfied: cbgen>=1.0.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (1.0.5)\n",
      "Requirement already satisfied: requests>=1.25.8 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (2.32.5)\n",
      "Requirement already satisfied: texttable>=1.6.2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from bgen-reader>=2.0.7->limix-lmm) (1.7.0)\n",
      "Requirement already satisfied: pooch in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from cbgen>=1.0.1->bgen-reader>=2.0.7->limix-lmm) (1.8.2)\n",
      "Requirement already satisfied: urllib3>=1.26 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from cbgen>=1.0.1->bgen-reader>=2.0.7->limix-lmm) (2.5.0)\n",
      "Requirement already satisfied: pycparser in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from cffi>=1.17->pandas-plink) (2.22)\n",
      "Requirement already satisfied: click>=8.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (8.3.0)\n",
      "Requirement already satisfied: cloudpickle>=3.0.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (3.1.2)\n",
      "Requirement already satisfied: fsspec>=2021.09.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (2025.10.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (25.0)\n",
      "Requirement already satisfied: partd>=1.4.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (1.4.2)\n",
      "Requirement already satisfied: pyyaml>=5.3.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (6.0.3)\n",
      "Requirement already satisfied: toolz>=0.10.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (1.1.0)\n",
      "Requirement already satisfied: importlib-metadata>=4.13.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[dataframe]->pandas-plink) (8.7.0)\n",
      "Requirement already satisfied: dask-expr<1.2,>=1.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask[array,bag,dataframe,delayed]>=2.12.0->bgen-reader>=2.0.7->limix-lmm) (1.1.19)\n",
      "Requirement already satisfied: pyarrow>=14.0.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from dask-expr<1.2,>=1.1->dask[array,bag,dataframe,delayed]>=2.12.0->bgen-reader>=2.0.7->limix-lmm) (22.0.0)\n",
      "Requirement already satisfied: zipp>=3.20 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from importlib-metadata>=4.13.0->dask[dataframe]->pandas-plink) (3.23.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (1.3.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (4.60.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (1.4.9)\n",
      "Requirement already satisfied: pillow>=8 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (12.0.0)\n",
      "Requirement already satisfied: pyparsing>=3 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from matplotlib>=2.2.3->limix-lmm) (3.2.5)\n",
      "Requirement already satisfied: locket in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from partd>=1.4.0->dask[dataframe]->pandas-plink) (1.0.0)\n",
      "Requirement already satisfied: exceptiongroup>=1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (1.3.0)\n",
      "Requirement already satisfied: iniconfig>=1.0.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (2.3.0)\n",
      "Requirement already satisfied: pluggy<2,>=1.5 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (1.6.0)\n",
      "Requirement already satisfied: pygments>=2.7.2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (2.19.2)\n",
      "Requirement already satisfied: tomli>=1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pytest>=3.6.3->limix-lmm) (2.3.0)\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from exceptiongroup>=1->pytest>=3.6.3->limix-lmm) (4.15.0)\n",
      "Requirement already satisfied: six>=1.5 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas<3.0,>=2.2->pandas-plink) (1.17.0)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from requests>=1.25.8->bgen-reader>=2.0.7->limix-lmm) (3.4.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from requests>=1.25.8->bgen-reader>=2.0.7->limix-lmm) (3.11)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from requests>=1.25.8->bgen-reader>=2.0.7->limix-lmm) (2025.11.12)\n",
      "Requirement already satisfied: scikit-learn>=0.20.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from geno-sugar->limix-lmm) (1.7.2)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from scikit-learn>=0.20.0->geno-sugar->limix-lmm) (1.5.2)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from scikit-learn>=0.20.0->geno-sugar->limix-lmm) (3.6.0)\n",
      "Requirement already satisfied: pydantic in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (2.12.5)\n",
      "Requirement already satisfied: typeguard in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (4.4.4)\n",
      "Requirement already satisfied: typing_inspect>=0.6.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (0.9.0)\n",
      "Requirement already satisfied: pandas-stubs in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (2.3.2.250926)\n",
      "Requirement already satisfied: scipy-stubs in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (1.15.3.0)\n",
      "Requirement already satisfied: distributed in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandera[dask,mypy]->pandas-plink) (2024.11.2)\n",
      "Requirement already satisfied: mypy-extensions>=0.3.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from typing_inspect>=0.6.0->pandera[dask,mypy]->pandas-plink) (1.1.0)\n",
      "Requirement already satisfied: jinja2>=2.10.3 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (3.0.3)\n",
      "Requirement already satisfied: msgpack>=1.0.2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (1.1.2)\n",
      "Requirement already satisfied: psutil>=5.8.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (7.1.3)\n",
      "Requirement already satisfied: sortedcontainers>=2.0.5 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (2.4.0)\n",
      "Requirement already satisfied: tblib>=1.6.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (3.2.2)\n",
      "Requirement already satisfied: tornado>=6.2.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (6.5.2)\n",
      "Requirement already satisfied: zict>=3.0.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from distributed->pandera[dask,mypy]->pandas-plink) (3.0.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from jinja2>=2.10.3->distributed->pandera[dask,mypy]->pandas-plink) (3.0.3)\n",
      "Requirement already satisfied: types-pytz>=2022.1.1 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pandas-stubs->pandera[dask,mypy]->pandas-plink) (2025.2.0.20251108)\n",
      "Requirement already satisfied: platformdirs>=2.5.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pooch->cbgen>=1.0.1->bgen-reader>=2.0.7->limix-lmm) (4.5.0)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pydantic->pandera[dask,mypy]->pandas-plink) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.41.5 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pydantic->pandera[dask,mypy]->pandas-plink) (2.41.5)\n",
      "Requirement already satisfied: typing-inspection>=0.4.2 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from pydantic->pandera[dask,mypy]->pandas-plink) (0.4.2)\n",
      "Requirement already satisfied: optype>=0.9.3 in /opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages (from scipy-stubs->pandera[dask,mypy]->pandas-plink) (0.9.3)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.3\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m26.0.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install pandas-plink limix-lmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "56aa0761",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from pandas_plink import read_plink\n",
    "from limix_lmm import LMM\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scipy.stats as st\n",
    "import statsmodels.api as sm\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from tqdm import tqdm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2bd7a5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "import scipy.stats as st\n",
    "\n",
    "if not hasattr(scipy, 'dot'):\n",
    "    scipy.dot = np.dot\n",
    "if not hasattr(scipy, 'einsum'):\n",
    "    scipy.einsum = np.einsum\n",
    "if not hasattr(scipy, 'log'):\n",
    "    scipy.log = np.log\n",
    "if not hasattr(scipy, 'sign'):\n",
    "    scipy.sign = np.sign\n",
    "if not hasattr(scipy, 'sqrt'):\n",
    "    scipy.sqrt = np.sqrt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29dbe785",
   "metadata": {},
   "source": [
    "### Preprocessing pipeline\n",
    "\n",
    "BFILE=/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/ALL.chr22_GRCh38.genotypes.20170504\n",
    "\n",
    "OUTDIR=/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/qc\n",
    "\n",
    "mkdir -p \"$OUTDIR\"\n",
    "\n",
    "\n",
    "plink --bfile \"$BFILE\" --freq --missing --out \"$OUTDIR\"/initial_qc\n",
    "\n",
    "\n",
    "Preprocessing \n",
    "\n",
    "We only kept SNP like variants (single base substitutions), dropped variants with missing rate > 0.02 and minor allele frequency < 0.01 to keep only commons, and also variants that have extreme deviations that are caused by noise rather than signal\n",
    "\n",
    "(filtered data set written as a new bed/bim/fam)\n",
    "\n",
    "\n",
    "plink --bfile \"$BFILE\" \\\n",
    "  --snps-only \\\n",
    "  --geno 0.02 \\\n",
    "  --maf 0.01 \\\n",
    "  --hwe 1e-6 \\\n",
    "  --make-bed \\\n",
    "  --out \"$OUTDIR\"/chr22_step1_common\n",
    "\n",
    "\n",
    "check if there are duplicate variant ids\n",
    "\n",
    "plink --bfile \"$OUTDIR\"/chr22_step1_common \\\n",
    "  --list-duplicate-vars ids-only suppress-first \\\n",
    "  --out \"$OUTDIR\"/dupcheck\n",
    "\n",
    "wc -l \"$OUTDIR\"/dupcheck.dupvar\n",
    "\n",
    "there were no duplicates (0)\n",
    "\n",
    "plink --bfile \"$OUTDIR\"/chr22_step1_common \\\n",
    "  --exclude \"$OUTDIR\"/dupcheck.dupvar \\\n",
    "  --make-bed \\\n",
    "  --out \"$OUTDIR\"/chr22_step2_nodup\n",
    "\n",
    "sample level missingness, none was filtered (both chr22_step2_nodup.fam and chr22_step3_sampleqc.fam are same size 2504)\n",
    "\n",
    "plink --bfile \"$OUTDIR\"/chr22_step2_nodup \\\n",
    "  --mind 0.02 \\\n",
    "  --make-bed \\\n",
    "  --out \"$OUTDIR\"/chr22_step3_sampleqc\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb4e96f0",
   "metadata": {},
   "source": [
    "INITIAL:\n",
    "\n",
    "Number of variants: 109827 \n",
    "\n",
    "Number of samples: 2504\n",
    "\n",
    "AFTER QC:\n",
    "\n",
    "Number of variants: 59743\n",
    "\n",
    "Number of samples: 2504\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e77b22dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 47.76it/s]\n",
      "/opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages/pandera/_pandas_deprecated.py:146: FutureWarning: Importing pandas-specific classes and functions from the\n",
      "top-level pandera module will be **removed in a future version of pandera**.\n",
      "If you're using pandera to validate pandas objects, we highly recommend updating\n",
      "your import:\n",
      "\n",
      "```\n",
      "# old import\n",
      "import pandera as pa\n",
      "\n",
      "# new import\n",
      "import pandera.pandas as pa\n",
      "```\n",
      "\n",
      "If you're using pandera to validate objects from other compatible libraries\n",
      "like pyspark or polars, see the supported libraries section of the documentation\n",
      "for more information on how to import pandera:\n",
      "\n",
      "https://pandera.readthedocs.io/en/stable/supported_libraries.html\n",
      "\n",
      "To disable this warning, set the environment variable:\n",
      "\n",
      "```\n",
      "export DISABLE_PANDERA_IMPORT_WARNING=True\n",
      "```\n",
      "\n",
      "  warnings.warn(_future_warning, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "bfile = 'data/qc/chr22_step3_sampleqc'\n",
    "bim, fam, G = read_plink(bfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "efab13a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_real = G.compute().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b132609a",
   "metadata": {},
   "source": [
    "Standardize the genotype matrix, so that all SNPs are on the same scale, 3 variants were dropped, which had a standard deviation equals to 0 among all individuals. (bim annotation table must be also updated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4e84ae63",
   "metadata": {},
   "outputs": [],
   "source": [
    "#whole chromosome 22\n",
    "\n",
    "mu_full = X_real.mean(axis=0)\n",
    "sd_full = X_real.std(axis=0, ddof=0)\n",
    "keep_full = sd_full > 1e-12\n",
    "keep_idx  = np.where(keep_full)[0] \n",
    "standardized_X = (X_real[:, keep_full] - mu_full[keep_full]) / sd_full[keep_full]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e689eb40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bim data for variants that werent dropped after the standardization\n",
    "bim_kept = bim.iloc[keep_idx].copy().reset_index(drop=True)\n",
    "\n",
    "bim_kept[\"orig_bim_idx\"] = keep_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "339425d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2504, 59740)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "standardized_X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8bf98ece",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_pheno(X, idx_caus, var_expl, rng, direction=None):\n",
    "    # Ensure that the number of causal variant indices matches the number of variances explained.\n",
    "    assert len(idx_caus) == len(var_expl)\n",
    "\n",
    "    # If no direction is provided, randomly assign a positive or negative direction for each causal variant.\n",
    "    if direction is None:\n",
    "        direction = 2. * (rng.random(len(idx_caus)) > 0.5) - 1.\n",
    "    # Ensure that the number of directions matches the number of causal variant indices.\n",
    "    assert len(idx_caus) == len(direction)\n",
    "\n",
    "    # Compute the remaining variance after accounting for the variance explained by the causal variants.\n",
    "    ve = 1 - var_expl.sum()\n",
    "    # Ensure that the total variance explained by causal variants is less than 1.\n",
    "    assert ve > 0, 'sum(var_expl) should be < 1'\n",
    "\n",
    "    # Compute the effect sizes for the causal variants based on the variance they explain and their direction.\n",
    "    beta = np.sqrt(var_expl) * direction\n",
    "\n",
    "    # Extract the columns of X corresponding to the causal variants and standardize them.\n",
    "    Xc = X[:, idx_caus]\n",
    "    Xc = (Xc - Xc.mean(0)) / Xc.std(0)\n",
    "\n",
    "    # Compute the genetic component of the phenotype.\n",
    "    yg = Xc.dot(beta)[:, None]\n",
    "    # Compute the noise component of the phenotype.\n",
    "    yn = np.sqrt(ve) * rng.standard_normal((X.shape[0], 1))\n",
    "\n",
    "    # Sum the genetic and noise components to get the simulated phenotype.\n",
    "    y = yg + yn\n",
    "\n",
    "    # Initialize the real effect sizes for all variants in X as zeros.\n",
    "    beta_real = np.zeros(X.shape[1])\n",
    "    # Update the real effect sizes for the causal variants.\n",
    "    beta_real[idx_caus] = beta\n",
    "\n",
    "    # Standardize the phenotypic values to have mean 0 and standard deviation 1.\n",
    "    ystd = y.std()\n",
    "    y = (y - y.mean()) / ystd\n",
    "    # Adjust the real effect sizes accordingly after standardizing y.\n",
    "    beta_real = beta_real / ystd\n",
    "\n",
    "    return y, beta_real"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "58050ba8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def qq_plot(p_values, title):\n",
    "    \"\"\"\n",
    "    Create a QQ plot given a list of p-values.\n",
    "\n",
    "    Parameters:\n",
    "    - p_values: list of p-values\n",
    "    - title: title for the plot\n",
    "    \"\"\"\n",
    "\n",
    "    # Sort p-values\n",
    "    observed = -np.log10(np.sort(p_values))\n",
    "    expected = -np.log10(np.arange(1, len(p_values) + 1) / (len(p_values) + 2))\n",
    "\n",
    "    # Create the QQ plot\n",
    "    plt.scatter(expected, observed, marker='.')\n",
    "    plt.plot([0, max(expected)], [0, max(expected)], color='red', linestyle='--')\n",
    "    plt.xlabel('Expected -log10(P-value)')\n",
    "    plt.ylabel('Observed -log10(P-value)')\n",
    "    plt.title(title)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "056e031b",
   "metadata": {},
   "outputs": [],
   "source": [
    "m = np.arange(10, 101, 10)\n",
    "heritability_vals = np.linspace(0.1, 0.6, 6).astype(np.float64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1929de43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6a926483",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "base_dir = Path(\"/Users/oykusuoglu/gobi/gobi_gwas/oyku/plots\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3e451c5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages/limix_lmm/lmm_core.py:322: RuntimeWarning: divide by zero encountered in divide\n",
      "  ste = beta / z\n"
     ]
    }
   ],
   "source": [
    "for heritability in heritability_vals:\n",
    "    for count_causal in m:\n",
    "        # Phenotype simulation\n",
    "        rng = np.random.default_rng(42)\n",
    "        idx_caus_loop = rng.choice(standardized_X.shape[1], size=count_causal, replace=False)\n",
    "        var_expl_loop = np.repeat(heritability/count_causal, count_causal)\n",
    "\n",
    "        y_loop, beta_real_loop = simulate_pheno(standardized_X, idx_caus_loop, var_expl_loop, rng)\n",
    "\n",
    "        # Merging PCs with phenotype\n",
    "        pheno = fam[[\"fid\",\"iid\"]].copy()\n",
    "        pheno.columns = [\"FID\",\"IID\"]\n",
    "        pheno[\"y\"] = y_loop.reshape(-1)\n",
    "\n",
    "        pcs = pd.read_csv(\"data/pca/chr22_pca.eigenvec\", sep=r\"\\s+\", header=None, engine=\"python\")\n",
    "        pcs.columns = [\"FID\",\"IID\"] + [f\"PC{i}\" for i in range(1, pcs.shape[1]-1)]\n",
    "\n",
    "        df = pheno.merge(pcs, on=[\"FID\",\"IID\"], how=\"inner\", validate=\"one_to_one\")\n",
    "        k = 10\n",
    "        F = np.column_stack([np.ones((df.shape[0], 1)), df[[f\"PC{i}\" for i in range(1, k+1)]].to_numpy()])\n",
    "\n",
    "        # Training split\n",
    "        X_train, X_test, y_train, y_test, F_train, F_test = train_test_split(\n",
    "                                                            standardized_X, y_loop, F,\n",
    "                                                            test_size=0.3,\n",
    "                                                            random_state=42,\n",
    "                                                            shuffle=True\n",
    "                                                                        )\n",
    "        # GWAS\n",
    "        lmm = LMM(y_train, F_train)\n",
    "        lmm.process(X_train)\n",
    "        pv = lmm.getPv()\n",
    "        beta = lmm.getBetaSNP()\n",
    "        beta_ste = lmm.getBetaSNPste()\n",
    "\n",
    "        outdir = base_dir / f\"h2_{heritability}_causal_vars{count_causal}\"\n",
    "        outdir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "        # plotting\n",
    "        fig = plt.figure()\n",
    "        qq_plot(pv, f'QQ Plot for heritability: {heritability} number of causal variants: {count_causal}')\n",
    "        fig.tight_layout()\n",
    "        fig.savefig(outdir / f\"qq_h2_{heritability}_causal_vars{count_causal}.png\", dpi=300, bbox_inches=\"tight\")\n",
    "        plt.close(fig)\n",
    "\n",
    "        x = bim_kept['pos'].values\n",
    "        fig2 = plt.figure(figsize=(10, 6))\n",
    "        ax1 = fig2.add_subplot(2, 1, 1)\n",
    "        ax2 = fig2.add_subplot(2, 1, 2, sharex=ax1)\n",
    "\n",
    "        ax1.set_title(f\"Real effect size for heritability: {heritability} number of causal variants: {count_causal}\")\n",
    "        ax1.plot(x, beta_real_loop, \".k\")\n",
    "        ax1.set_ylabel(\"eff size\")\n",
    "\n",
    "        ax2.set_title(f\"GWAS results for heritability: {heritability} number of causal variants: {count_causal}\")\n",
    "        ax2.plot(x, -np.log10(pv), \".k\")\n",
    "        ax2.set_ylabel(r\"-log$_{10}$ P\")\n",
    "        ax2.set_xlabel(\"position\")\n",
    "\n",
    "        fig2.tight_layout()\n",
    "        fig2.savefig(outdir / f\"effects_gwas_h2{heritability}_causal_vars{count_causal}.png\",\n",
    "                    dpi=300, bbox_inches=\"tight\")\n",
    "        plt.close(fig2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d73706d",
   "metadata": {},
   "source": [
    "### Phase 2: GWAS / Feature Selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e49ca3e5",
   "metadata": {},
   "source": [
    "#### 3. Clumping using PLINK "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63dedee4",
   "metadata": {},
   "source": [
    "First create a file with SNP ids and their corresponding p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a2443f92",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_for_clump = pd.DataFrame({\n",
    "    \"SNP\": bim_kept[\"snp\"].astype(str),   \n",
    "    \"P\": np.asarray(pv, dtype=float)\n",
    "})\n",
    "gwas_for_clump.to_csv(\"gwas_for_clump.txt\", sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gobi_venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
