{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0237a8e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from pandas_plink import read_plink\n",
    "from limix_lmm import LMM\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scipy.stats as st\n",
    "import statsmodels.api as sm\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from tqdm import tqdm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "083fee64",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "import scipy.stats as st\n",
    "\n",
    "if not hasattr(scipy, 'dot'):\n",
    "    scipy.dot = np.dot\n",
    "if not hasattr(scipy, 'einsum'):\n",
    "    scipy.einsum = np.einsum\n",
    "if not hasattr(scipy, 'log'):\n",
    "    scipy.log = np.log\n",
    "if not hasattr(scipy, 'sign'):\n",
    "    scipy.sign = np.sign\n",
    "if not hasattr(scipy, 'sqrt'):\n",
    "    scipy.sqrt = np.sqrt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "147c7dc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.9.0-b.7.11 64-bit (19 Aug 2025)           cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_prune.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/raw/ALL.chr22_GRCh38.genotypes.20170504\n",
      "  --indep-pairwise 200 50 0.2\n",
      "  --out /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_prune\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "109827 variants loaded from .bim file.\n",
      "2504 people (0 males, 0 females, 2504 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_prune.nosex\n",
      ".\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2504 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.999911.\n",
      "109827 variants and 2504 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "Pruned 102281 variants from chromosome 22, leaving 7546.\n",
      "Pruning complete.  102281 of 109827 variants removed.\n",
      "Marker lists written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_prune.prune.in\n",
      "and\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_prune.prune.out\n",
      ".\n",
      "PLINK v1.9.0-b.7.11 64-bit (19 Aug 2025)           cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_pca10.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/raw/ALL.chr22_GRCh38.genotypes.20170504\n",
      "  --extract /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_prune.prune.in\n",
      "  --out /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_pca10\n",
      "  --pca 10\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "109827 variants loaded from .bim file.\n",
      "2504 people (0 males, 0 females, 2504 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_pca10.nosex\n",
      ".\n",
      "--extract: 7552 variants remaining.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: At least 2 duplicate IDs in --extract file.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using up to 9 threads (change this with --threads).\n",
      "Before main variant filters, 2504 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.999756.\n",
      "7552 variants and 2504 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "Relationship matrix calculation complete.\n",
      "--pca: Results saved to and eigenvectors]\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_pca10.eigenval\n",
      "and\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_pca10.eigenvec\n",
      ".\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -euo pipefail\n",
    "\n",
    "RAW_BFILE=\"/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/raw/ALL.chr22_GRCh38.genotypes.20170504\"\n",
    "OUTDIR_PC=\"/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc\"\n",
    "\n",
    "mkdir -p \"$OUTDIR_PC\"\n",
    "\n",
    "plink --bfile \"$RAW_BFILE\" \\\n",
    "  --indep-pairwise 200 50 0.2 \\\n",
    "  --out \"$OUTDIR_PC/chr22_raw_prune\"\n",
    "\n",
    "#### PCA on pruned RAW variants (writes: .eigenvec, .eigenval)\n",
    "plink --bfile \"$RAW_BFILE\" \\\n",
    "  --extract \"$OUTDIR_PC/chr22_raw_prune.prune.in\" \\\n",
    "  --pca 10 \\\n",
    "  --out \"$OUTDIR_PC/chr22_raw_pca10\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6aa8b0ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.9.0-b.7.11 64-bit (19 Aug 2025)           cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/initial_qc.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/raw/ALL.chr22_GRCh38.genotypes.20170504\n",
      "  --freq\n",
      "  --missing\n",
      "  --out /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/initial_qc\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "109827 variants loaded from .bim file.\n",
      "2504 people (0 males, 0 females, 2504 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/initial_qc.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2504 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.999911.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/initial_qc.frq .\n",
      "--missing: Sample missing data report written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/initial_qc.imiss,\n",
      "and variant-based missing data report written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/initial_qc.lmiss.\n",
      "PLINK v1.9.0-b.7.11 64-bit (19 Aug 2025)           cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step1_common.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/raw/ALL.chr22_GRCh38.genotypes.20170504\n",
      "  --geno 0.02\n",
      "  --hwe 1e-6\n",
      "  --maf 0.01\n",
      "  --make-bed\n",
      "  --out /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step1_common\n",
      "  --snps-only\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "97000 out of 109827 variants loaded from .bim file.\n",
      "2504 people (0 males, 0 females, 2504 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step1_common.nosex\n",
      ".\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2504 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.999964.\n",
      "7 variants removed due to missing genotype data (--geno).\n",
      "--hwe: 37250 variants removed due to Hardy-Weinberg exact test.\n",
      "0 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "59743 variants and 2504 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step1_common.bed\n",
      "+\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step1_common.bim\n",
      "+\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step1_common.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.9.0-b.7.11 64-bit (19 Aug 2025)           cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/dupcheck.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step1_common\n",
      "  --list-duplicate-vars ids-only suppress-first\n",
      "  --out /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/dupcheck\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "59743 variants loaded from .bim file.\n",
      "2504 people (0 males, 0 females, 2504 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/dupcheck.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2504 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is in [0.9999995, 1).\n",
      "59743 variants and 2504 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--list-duplicate-vars report written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/dupcheck.dupvar .\n",
      "       0 /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/dupcheck.dupvar\n",
      "PLINK v1.9.0-b.7.11 64-bit (19 Aug 2025)           cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step2_nodup.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step1_common\n",
      "  --exclude /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/dupcheck.dupvar\n",
      "  --make-bed\n",
      "  --out /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step2_nodup\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "59743 variants loaded from .bim file.\n",
      "2504 people (0 males, 0 females, 2504 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step2_nodup.nosex\n",
      ".\n",
      "--exclude: 59743 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2504 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is in [0.9999995, 1).\n",
      "59743 variants and 2504 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step2_nodup.bed\n",
      "+\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step2_nodup.bim\n",
      "+\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step2_nodup.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.9.0-b.7.11 64-bit (19 Aug 2025)           cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step3_sampleqc.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step2_nodup\n",
      "  --make-bed\n",
      "  --mind 0.02\n",
      "  --out /Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step3_sampleqc\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "59743 variants loaded from .bim file.\n",
      "2504 people (0 males, 0 females, 2504 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step3_sampleqc.nosex\n",
      ".\n",
      "0 people removed due to missing genotype data (--mind).\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2504 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is in [0.9999995, 1).\n",
      "59743 variants and 2504 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step3_sampleqc.bed\n",
      "+\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step3_sampleqc.bim\n",
      "+\n",
      "/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step3_sampleqc.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -euo pipefail\n",
    "\n",
    "RAW_BFILE=\"/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/raw/ALL.chr22_GRCh38.genotypes.20170504\"\n",
    "OUTDIR_QC=\"/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc\"\n",
    "\n",
    "mkdir -p \"$OUTDIR_QC\"\n",
    "\n",
    "#### Initial QC summaries on RAW\n",
    "plink --bfile \"$RAW_BFILE\" \\\n",
    "  --freq --missing \\\n",
    "  --out \"$OUTDIR_QC/initial_qc\"\n",
    "\n",
    "#### Variant-level filtering (SNPs only, geno, maf, hwe) -> new BED/BIM/FAM\n",
    "plink --bfile \"$RAW_BFILE\" \\\n",
    "  --snps-only \\\n",
    "  --geno 0.02 \\\n",
    "  --maf 0.01 \\\n",
    "  --hwe 1e-6 \\\n",
    "  --make-bed \\\n",
    "  --out \"$OUTDIR_QC/chr22_step1_common\"\n",
    "\n",
    "#### Duplicate variant ID check\n",
    "plink --bfile \"$OUTDIR_QC/chr22_step1_common\" \\\n",
    "  --list-duplicate-vars ids-only suppress-first \\\n",
    "  --out \"$OUTDIR_QC/dupcheck\"\n",
    "\n",
    "wc -l \"$OUTDIR_QC/dupcheck.dupvar\"\n",
    "\n",
    "#### Remove duplicates (safe even if file is empty; plink will just exclude none)\n",
    "plink --bfile \"$OUTDIR_QC/chr22_step1_common\" \\\n",
    "  --exclude \"$OUTDIR_QC/dupcheck.dupvar\" \\\n",
    "  --make-bed \\\n",
    "  --out \"$OUTDIR_QC/chr22_step2_nodup\"\n",
    "\n",
    "#### Sample-level missingness\n",
    "plink --bfile \"$OUTDIR_QC/chr22_step2_nodup\" \\\n",
    "  --mind 0.02 \\\n",
    "  --make-bed \\\n",
    "  --out \"$OUTDIR_QC/chr22_step3_sampleqc\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59c8dade",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "set -euo pipefail\n",
    "\n",
    "OUTDIR_QC=\"/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc\"\n",
    "OUTDIR_PC=\"/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc\"\n",
    "\n",
    "# IDs in final QC dataset\n",
    "cut -d' ' -f1,2 \"$OUTDIR_QC/chr22_step3_sampleqc.fam\" | sort > \"$OUTDIR_QC/final_ids.txt\"\n",
    "\n",
    "# IDs in PCA eigenvec\n",
    "awk '{print $1, $2}' \"$OUTDIR_PC/chr22_raw_pca10.eigenvec\" | sort > \"$OUTDIR_PC/pca_ids.txt\"\n",
    "\n",
    "# show any samples present in final but missing in PCA \n",
    "comm -23 \"$OUTDIR_QC/final_ids.txt\" \"$OUTDIR_PC/pca_ids.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85bbf7a8",
   "metadata": {},
   "source": [
    "Standardize PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a89e7b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_pca10.eigenvec.zscore'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "eigenvec = \"/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/chr22_raw_pca10.eigenvec\"\n",
    "\n",
    "df = pd.read_csv(eigenvec, sep=r\"\\s+\", header=None)\n",
    "ncols = df.shape[1]\n",
    "pc_cols = list(range(2, ncols))  # columns 2..end are PCs\n",
    "\n",
    "# z-score each PC across individuals\n",
    "pcs = df.iloc[:, pc_cols]\n",
    "pcs_z = (pcs - pcs.mean(axis=0)) / pcs.std(axis=0, ddof=0)\n",
    "\n",
    "df_z = df.copy()\n",
    "df_z.iloc[:, pc_cols] = pcs_z\n",
    "\n",
    "df_z.columns = [\"FID\", \"IID\"] + [f\"PC{i}\" for i in range(1, len(pc_cols)+1)]\n",
    "\n",
    "out = eigenvec.replace(\".eigenvec\", \".eigenvec.zscore\")\n",
    "df_z.to_csv(out, sep=\"\\t\", index=False)\n",
    "\n",
    "out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "373caace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(PC1     0.0\n",
       " PC2    -0.0\n",
       " PC3     0.0\n",
       " PC4     0.0\n",
       " PC5     0.0\n",
       " PC6    -0.0\n",
       " PC7    -0.0\n",
       " PC8     0.0\n",
       " PC9     0.0\n",
       " PC10    0.0\n",
       " dtype: float64,\n",
       " PC1     1.0\n",
       " PC2     1.0\n",
       " PC3     1.0\n",
       " PC4     1.0\n",
       " PC5     1.0\n",
       " PC6     1.0\n",
       " PC7     1.0\n",
       " PC8     1.0\n",
       " PC9     1.0\n",
       " PC10    1.0\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_z[[f\"PC{i}\" for i in range(1, 11)]].mean().round(6), df_z[[f\"PC{i}\" for i in range(1, 11)]].std(ddof=0).round(6)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e1f275",
   "metadata": {},
   "source": [
    "Check how data looks like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "baf0847f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 39.29it/s]\n",
      "/opt/anaconda3/envs/gobi_venv/lib/python3.10/site-packages/pandera/_pandas_deprecated.py:146: FutureWarning: Importing pandas-specific classes and functions from the\n",
      "top-level pandera module will be **removed in a future version of pandera**.\n",
      "If you're using pandera to validate pandas objects, we highly recommend updating\n",
      "your import:\n",
      "\n",
      "```\n",
      "# old import\n",
      "import pandera as pa\n",
      "\n",
      "# new import\n",
      "import pandera.pandas as pa\n",
      "```\n",
      "\n",
      "If you're using pandera to validate objects from other compatible libraries\n",
      "like pyspark or polars, see the supported libraries section of the documentation\n",
      "for more information on how to import pandera:\n",
      "\n",
      "https://pandera.readthedocs.io/en/stable/supported_libraries.html\n",
      "\n",
      "To disable this warning, set the environment variable:\n",
      "\n",
      "```\n",
      "export DISABLE_PANDERA_IMPORT_WARNING=True\n",
      "```\n",
      "\n",
      "  warnings.warn(_future_warning, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "bfile = '/Users/oykusuoglu/gobi/gobi_gwas/oyku/data/pca/before_qc/qc/chr22_step3_sampleqc'\n",
    "bim, fam, G = read_plink(bfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4116f0f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[2., 2., 2., ..., 2., 1., 2.],\n",
       "       [1., 2., 2., ..., 2., 1., 2.],\n",
       "       [2., 2., 2., ..., 1., 2., 2.],\n",
       "       ...,\n",
       "       [2., 2., 2., ..., 2., 2., 2.],\n",
       "       [2., 2., 1., ..., 1., 2., 2.],\n",
       "       [1., 2., 1., ..., 1., 2., 2.]], dtype=float32)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_real = G.compute().T\n",
    "X_real"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gobi_venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
